shared/constants/fs.js

```diff
@@ -264,3 +282,140 @@ export const isImage = (name: string): boolean => {
+    if (!folder.problem_set) {
+      return {
+        ...folder,
+        isIgnored,
+        isNew,
+        needsRekey: false,
+      }
+    }
+
+    let waitingForParticipantUnlock = []
+    let youCanUnlock = []
+    let needsRekey = false
+    const solutions = folder.problem_set.solution_kids || {}
+
+    if (Object.keys(solutions).length) {
+      needsRekey = true
+    }
+
+    if (folder.problem_set.can_self_help) {
+      const mySolutions = solutions[myKID] || []
+      youCanUnlock = mySolutions.map(kid => {
+        const device = favoritesResult.devices[kid]
+        return {...device, deviceID: kid}
+      })
+    } else {
+      waitingForParticipantUnlock = Object.keys(solutions).map(userID => {
+        const devices = solutions[userID].map(kid => favoritesResult.devices[kid].name)
+        const numDevices = devices.length
+        const last = numDevices > 1 ? devices.pop() : null
+
+        return {
+          name: favoritesResult.users[userID],
+          devices: `Tell them to turn on${numDevices > 1 ? ':' : ' '} ${devices.join(', ')}${
+            last ? ` or ${last}` : ''
+          }.`,
+        }
+      })
+    }
+    return {
+      ...folder,
+      waitingForParticipantUnlock,
+      youCanUnlock,
+      needsRekey,
+      isIgnored,
+      isNew,
+    }
+  }
+
+  return [
+    ...favoritesResult.favorites.map(mapFolderWithMeta({isIgnored: false, isNew: false})),
+    ...favoritesResult.ignored.map(mapFolderWithMeta({isIgnored: true, isNew: false})),
+    ...favoritesResult.new.map(mapFolderWithMeta({isIgnored: false, isNew: true})),
+  ]
+}
+
+export const folderToPathItems = (
+  txt: string = '',
+  username: string,
+  loggedIn: boolean
+): I.Map<Types.Path, Types.FavoriteItem> => {
+  let favoritesResult
+  let badges = {
+    '/keybase/private': 0,
+    '/keybase/public': 0,
+    '/keybase/team': 0,
+  }
+  let favoriteChildren = {
+    '/keybase/private': new Set(),
+    '/keybase/public': new Set(),
+    '/keybase/team': new Set(),
+  }
+  try {
+    favoritesResult = JSON.parse(txt)
+  } catch (err) {
+    logger.warn('Invalid json from getFavorites: ', err)
+    return I.Map()
+  }
+
+  const myKID = findKey(favoritesResult.users, name => name === username)
+
+  // figure out who can solve the rekey
+  const folders: Array<Types.FolderRPCWithMeta> = _fillMetadataInFavoritesResult(favoritesResult, myKID)
+  const favoriteFolders = folders.map(
+    ({name, folderType, isIgnored, isNew, needsRekey, waitingForParticipantUnlock, youCanUnlock}) => {
+      const folderTypeString = FolderTypeToString(folderType)
+      const folderParent = `/keybase/${folderTypeString}`
+      const preferredName = tlfToPreferredOrder(name, username)
+      const folderPathString = `${folderParent}/${preferredName}`
+      const folderPath = Types.stringToPath(folderPathString)
+      favoriteChildren[folderParent].add(preferredName)
+      if (isNew) {
+        badges[folderParent] += 1
+      }
```

Comment:
If we have a separate iteration on `favoriteFolders` for `{isNew}`, we can make the `folders.map()` declarative. It's an additional iteration but it's clearer on first glance what this does. I.e. something like:

```
folders.filter({isNew} => isNew).forEach({folderParent} => {
  badges[folderParent] += 1
})
const favoriteFolders = folders.map( ({ .... }) => ({
....
}) )
```

Reply:
But then we have a separate iteration, which isn't cache-friendly.

Reply:
And we don't store the folder parents, so we'd have to calculate each one anyway. I'd say we should leave it.

Reply:
> cache-friendly

I feel the `folders` should be small enough that this doesn't matter. But I'm OK with leaving this too.